internal/wasm/table.go

```diff
@@ -18,68 +24,197 @@ type ElementSegment struct {
 	Init []Index
 }
 
-// TableInstance represents a table instance in a store.
+// TableInstance represents a table of (ElemTypeFuncref) elements in a module.
+//
 // See https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#table-instances%E2%91%A0
 type TableInstance struct {
-	// Table holds the table elements managed by this table instance.
-	Table []uintptr
+	// Table holds the Engine-specific compiled functions mapped by element index.
+	Table []interface{}
 
-	// Min is the minimum (function) elements in this table.
+	// Min is the minimum (function) elements in this table and cannot grow to accommodate ElementSegment.
 	Min uint32
 
 	// Max if present is the maximum (function) elements in this table, or nil if unbounded.
 	Max *uint32
 }
 
-func (m *Module) buildTable() *TableInstance {
-	table := m.TableSection
-	if table != nil {
-		return &TableInstance{
-			Table: make([]uintptr, table.Min),
-			Min:   table.Min,
-			Max:   table.Max,
+// validatedElementSegment is like ElementSegment except the inputs are expanded and validated based on defining module.
+//
+// Note: The global imported at globalIdx may have an offset value that is out-of-bounds for the corresponding table.
+type validatedElementSegment struct {
+	// opcode is OpcodeGlobalGet or OpcodeI32Const
+	opcode Opcode
+
+	// arg0 is the arg to opcode, which results in the offset to add to init indices.
+	//  * OpcodeGlobalGet: position in the global index namespace of an imported Global ValueTypeI32 holding the offset.
+	//  * OpcodeI32Const: a constant ValueTypeI32 offset.
+	arg0 uint32
```

Comment:
I would drop `0` as in anyway there would be at most one arg

Reply:
sgtm!